Сарапин Илья

unmorsino opened the pull request.

Commit:
Делегировал создание MainForm контейнеру

Commit:
Делигировал создание IUiActions контейнеру

Commit:
Рефакторинг DI для KochFractalAction

Commit:
реорганизовал зависимости KochFractalAction

Commit:
добавление фабрики для DragonPainter

Commit:
добавил функциональную фабрику для DragonSettings

Commit:
перевел DragonPainter на цвета палитры

Commit:
удалил зависимость ImageSettingsAction от INeed

Commit:
избавился от INeed

Commit:
заменил ручной биндинг IUiAction в DI контейнере на автоматический

Commit:
Добавил в IUiAction поле Order для сортировки пунктов

maracuda commented on FractalPainter/App/Program.cs:
```diff
@@ -15,13 +21,39 @@ private static void Main()
             try
             {
                 var container = new StandardKernel();
+                container.Bind<MainForm>().ToSelf().InSingletonScope();
```
Эта строчка не нужна
Контейнер при наличии всех нужных зависимостей сам соберет форму

Однако в некоторых контейнерах это не так и прилетит исключение, если форма не будет зарегистрирована. О плюсах и минусах такого подхода можешь порассуждать+)

unmorsino replied:
Да, я привык к SimpleInjector, там по умолчанию нужно все регистрировать явно.
Собственно плюсы в том что простые классы регистрируются автомагически, проще писать, проще читать конфиг контейнера.
Минусы в том что можно забыть сконфигурировать что-то что действительно надо сконфигурировать, как получилось с DragonSettings

maracuda commented on FractalPainter/App/Program.cs:
```diff
@@ -15,13 +21,39 @@ private static void Main()
+                {
+                    var imageSettings = ctx.Kernel.Get<SettingsManager>().Load().ImageSettings;
+                    var pictureBox = new PictureBoxImageHolder();
+                    pictureBox.RecreateImage(imageSettings);
```
Этого здесь быть не должно
Похоже на бизнес-логику внутри конфигурации контейнера
Давай попробуем сделать по-другому

Вспомни что за ошибка здесь была и подумай как по-другому решить, прийди ко мне если сам не сможешь

maracuda commented on FractalPainter/App/Program.cs:
```diff
@@ -15,13 +21,39 @@ private static void Main()
+
+                container.Bind<Palette>().ToSelf().InSingletonScope();
+                
+                container.Bind<IImageHolder>().ToMethod(ctx => ctx.Kernel.Get<PictureBoxImageHolder>());
```
Это можно совместить с 37 строкой. У метода Bind есть перегрузка <T1, T2>

maracuda commented on FractalPainter/App/MainForm.cs:
```diff
@@ -1,55 +1,23 @@
-                    new ImageSettingsAction(),
-                    new PaletteSettingsAction()
-                })
+        public MainForm(IUiAction[] actions, SettingsManager settingsManager, PictureBoxImageHolder pictureBox, Palette palette)
```
Давай передадим сюда ImageSettings вместо менеджера, ведь это делает нас менее зависимыми от настроек

maracuda replied:
Кажется палитра здесь не нужна

maracuda commented on FractalPainter/App/Actions/DragonFractalAction.cs:
```diff
@@ -1,40 +1,40 @@
         {
-            var dragonSettings = CreateRandomSettings();
-            // редактируем настройки:
+            var dragonSettings = dragonSettingsFactory.Invoke();
```
Раньше при генерации настроек использовался new Random(), сейчас этого аргумента нет
Давай добавим и восстановим логику

Можно делать не Invoke, а просто factory()

unmorsino replied:
Поправил

Commit:
Вернул создание DragonSettings с помощью DragonSettingsGenerator

Commit:
удалил из конструктора Palette, которая не использовалась

Commit:
Изменил зависимость от SettingsManager на зависимость сразу от ImageSettings

Commit:
Исключил регистрацию классов на самих себя

Commit:
перенес логику конфигурации PictureBox из контейнера в MainForm,
В KochPainter перенес получение Size из ImageHolder на момент отрисовки, иначе получал исключение, что объект должен быть инициализирован

Commit:
Зарегистрировал зависимости PictureBoxImageHolder, IImageHolder через один обобщенный метод Bind

unmorsino reviewed (commented):
Исправил

Commit:
Перенес бизнес-логику по настройке DragonSettings из контейнера в DragonFractalAction
